packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+import { Slot, SlotProvider } from './slot'
+import { types } from '@mondrian-framework/model'
+import { functions } from '@mondrian-framework/module'
+
+//inspired by https://blog.cloudflare.com/counting-things-a-lot-of-different-things/
+
+/**
+ * Rate limit at `rateLimit` in the `samplingPeriod`.
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
+    this.rate = typeof rate === 'string' ? parseRate(rate) : rate
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
+    const slidingWindow = this.slidingWindows.get(key)
+    if (slidingWindow) {
+      return slidingWindow
+    }
+    const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
+    this.slidingWindows.set(key, newSlidingWindow)
+    return newSlidingWindow
+  }
+}
+
+export class SlidingWindow {
+  private readonly samplingPeriodSeconds: number
+  private readonly rateLimit: number
+  private readonly slots: Map<number, Slot> = new Map()
+  private readonly nowSeconds: () => number
+  private readonly slotProvider: SlotProvider
+  private readonly key: string
+  private rateLimitedUntil?: number
+
+  constructor({
+    rate,
+    nowSeconds,
+    slotProvider,
+    key,
+  }: {
+    rate: Rate
+    nowSeconds?: () => number
+    slotProvider: SlotProvider
+    key: string
+  }) {
+    if (rate.period < 1) {
+      throw new Error('Sampling period must be at least 1 second')
+    }
+    if (rate.requests < 0) {
+      throw new Error('Rate limit must be a positive duration')
+    }
+    this.samplingPeriodSeconds = rate.period * (rate.unit === 'hours' ? 3600 : rate.unit === 'minutes' ? 60 : 1)
+    this.rateLimit = rate.requests
+    this.nowSeconds = nowSeconds ?? (() => new Date().getTime() / 1000.0)
+    this.slotProvider = slotProvider
+    this.key = key
+  }
+
+  private getSlots(): [Slot, Slot] {
+    const now = this.nowSeconds()
+    const actualFrom = Math.floor(now / this.samplingPeriodSeconds) * this.samplingPeriodSeconds
+    const oldFrom = actualFrom - this.samplingPeriodSeconds
+
+    const oldSlot =
+      this.slots.get(oldFrom) ??
+      this.slotProvider.create({
+        fromSecond: oldFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(oldFrom, oldSlot)
+    const actualSlot =
+      this.slots.get(actualFrom) ??
+      this.slotProvider.create({
+        fromSecond: actualFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(actualFrom, actualSlot)
```

Comment:
Could you explain to me, in natural language, what's the purpose of this block of code?

Reply:
Get the two slot (old and actual)
<img width="797" alt="image" src="https://github.com/twinlogix/mondrian-framework/assets/50401517/90f5213c-2b29-4127-b72c-3f2968d7d485">
If they does not exists -> creates them.

After create a slot save it to the map `slots` so it can be retrieved later.

Reply:
Maybe it could be clearer to add another internal function `getOrCreate` to factor out these two pieces of code since they're kind of similar? Also it would be better at self-documenting its purpose